exporter/opencensusexporter/opencensus.go

```diff
@@ -129,22 +182,61 @@ func OpenCensusTraceExportersFromViper(v *viper.Viper) (tps []consumer.TraceCons
 	}
 
 	tps = append(tps, oexp)
+	doneFns = append(doneFns, oce.stop)
 
 	// TODO: (@odeke-em, @songya23) implement ExportMetrics for OpenCensus.
 	// mps = append(mps, oexp)
 	return
 }
 
+func (oce *ocagentExporter) stop() error {
+	wg := &sync.WaitGroup{}
+	var errors []error
+	var errorsMu sync.Mutex
+	visitedCnt := 0
+	for currExporter := range oce.exporters {
+		wg.Add(1)
+		go func(exporter *ocagent.Exporter) {
+			defer wg.Done()
+			err := exporter.Stop()
+			if err != nil {
+				errorsMu.Lock()
+				errors = append(errors, err)
+				errorsMu.Unlock()
+			}
+		}(currExporter)
+		visitedCnt++
+		if visitedCnt == cap(oce.exporters) {
+			// Visited and started Stop on all exporters, just wait for the stop to finish.
+			break
+		}
+	}
+
+	wg.Wait()
+	close(oce.exporters)
+
+	return internal.CombineErrors(errors)
+}
+
 func (oce *ocagentExporter) PushTraceData(ctx context.Context, td data.TraceData) (int, error) {
-	// Get an exporter worker round-robin
-	exporter := oce.exporters[atomic.AddUint32(&oce.counter, 1)%uint32(len(oce.exporters))]
+	// Get first available exporter.
+	exporter, ok := <-oce.exporters
+	if !ok {
+		err := &ocTraceExporterError{
+			code: errAlreadyStopped,
+			msg:  fmt.Sprintf("OpenCensus exporter was already stopped."),
+		}
+		return len(td.Spans), err
+	}
+
 	err := exporter.ExportTraceServiceRequest(
 		&agenttracepb.ExportTraceServiceRequest{
 			Spans:    td.Spans,
 			Resource: td.Resource,
 			Node:     td.Node,
 		},
 	)
+	oce.exporters <- exporter
```

Comment:
If I remember correctly writing on closed channel panics.

If you can guarantee that send() and PushTraceData() are never called concurrently that would be the best way instead of trying to make them safe to call concurrently, just add comments explaining that it is unsafe.

Reply:
that is correct, but the channel is only closed when all exporters were stopped at that moment there won't be any exporter in transition to be added back to the channel. If it was retrieved from the channel here it was not stopped yet so the channel was not closed yet.

Reply:
:+1: